// @progress/kendo-angular-toolbar/esm2015/toolbar.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ButtonsModule } from '@progress/kendo-angular-buttons';
import { PopupModule } from '@progress/kendo-angular-popup';
import { ResizeSensorModule } from '@progress/kendo-angular-common';
import { ToolBarComponent } from './toolbar.component';
import { ToolBarToolComponent } from './tools/toolbar-tool.component';
import { ToolBarButtonComponent } from './tools/toolbar-button.component';
import { ToolBarButtonGroupComponent } from './tools/toolbar-buttongroup.component';
import { ToolBarDropDownButtonComponent } from './tools/toolbar-dropdownbutton.component';
import { ToolBarSplitButtonComponent } from './tools/toolbar-splitbutton.component';
import { ToolBarSeparatorComponent } from './tools/toolbar-separator.component';
import { ToolBarSpacerComponent } from './tools/toolbar-spacer.component';
import { ToolBarRendererComponent } from './renderer.component';
import { ToolBarButtonListComponent } from './tools/toolbar-buttonlist.component';
import { LocalizedToolbarMessagesDirective } from './localization/localized-toolbar-messages.directive';
import { ToolbarCustomMessagesComponent } from './localization/custom-messages.component';
import * as i0 from "@angular/core";
const TOOLBAR_TOOLS = [
    ToolBarToolComponent,
    ToolBarButtonComponent,
    ToolBarButtonGroupComponent,
    ToolBarDropDownButtonComponent,
    ToolBarSplitButtonComponent,
    ToolBarSeparatorComponent,
    ToolBarSpacerComponent
];
const TOOLBAR_COMMON = [
    ToolBarRendererComponent,
    ToolBarButtonListComponent,
    ToolbarCustomMessagesComponent,
    LocalizedToolbarMessagesDirective
];
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi']) definition for the ToolBar component.
 *
 * The package exports:
 * - `ToolBarComponent`&mdash;The ToolBarComponent class.
 * - `ToolBarToolComponent`&mdash;The base Tool component class.
 * - `ToolBarButtonComponent`&mdash;The Button Tool component class.
 * - `ToolBarButtonGroupComponent`&mdash;The ButtonGroup Tool component class.
 * - `ToolBarDropDownButtonComponent`&mdash;The DropDownButton Tool component class.
 * - `ToolBarSplitButtonComponent`&mdash;The SplitButton Tool component class.
 * - `ToolBarSeparatorComponent`&mdash;The Separator Tool component class.
 */
export class ToolBarModule {
}
ToolBarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: ToolBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ToolBarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: ToolBarModule, declarations: [ToolBarComponent, ToolBarToolComponent,
        ToolBarButtonComponent,
        ToolBarButtonGroupComponent,
        ToolBarDropDownButtonComponent,
        ToolBarSplitButtonComponent,
        ToolBarSeparatorComponent,
        ToolBarSpacerComponent, ToolBarRendererComponent,
        ToolBarButtonListComponent,
        ToolbarCustomMessagesComponent,
        LocalizedToolbarMessagesDirective], imports: [CommonModule, ButtonsModule, PopupModule, ResizeSensorModule], exports: [ToolBarComponent, ToolBarToolComponent,
        ToolBarButtonComponent,
        ToolBarButtonGroupComponent,
        ToolBarDropDownButtonComponent,
        ToolBarSplitButtonComponent,
        ToolBarSeparatorComponent,
        ToolBarSpacerComponent, ToolbarCustomMessagesComponent, LocalizedToolbarMessagesDirective] });
ToolBarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: ToolBarModule, imports: [[CommonModule, ButtonsModule, PopupModule, ResizeSensorModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: ToolBarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ToolBarComponent, TOOLBAR_TOOLS, TOOLBAR_COMMON],
                    exports: [ToolBarComponent, TOOLBAR_TOOLS, ToolbarCustomMessagesComponent, LocalizedToolbarMessagesDirective],
                    imports: [CommonModule, ButtonsModule, PopupModule, ResizeSensorModule]
                }]
        }] });
