// @progress/kendo-angular-editor/esm2015/tools/shared/editor-command-button.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { EditorCommandBaseDirective } from './editor-command-base';
/**
 * @hidden
 */
export class EditorCommandButton extends EditorCommandBaseDirective {
    constructor(command, button, localization, providerService, toolsService) {
        super(command, button, localization, providerService, toolsService);
        this.command = command;
        this.button = button;
        this.localization = localization;
        this.providerService = providerService;
        this.toolsService = toolsService;
    }
    clickHandler() {
        this.editor.exec(this.command, this.editor.applyToWord);
        this.editor.focus();
    }
    pointerdownHandler(e) {
        e.preventDefault();
    }
    onStateChange(toolBarState) {
        this.button.selected = toolBarState[this.command].selected;
        this.button.disabled = toolBarState[this.command].disabled;
    }
}
