// @progress/kendo-angular-editor/esm2015/tools/shared/editor-command-dialog.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { EditorCommandBaseDirective } from './editor-command-base';
/**
 * @hidden
 */
export class EditorCommandDialog extends EditorCommandBaseDirective {
    constructor(dialog, button, localization, providerService, toolsService) {
        super(dialog, button, localization, providerService, toolsService);
        this.dialog = dialog;
        this.button = button;
        this.localization = localization;
        this.toolsService = toolsService;
    }
    clickHandler() {
        this.editor.openDialog(this.dialog);
    }
    onStateChange(toolBarState) {
        this.button.selected = toolBarState[this.command].selected;
        this.button.disabled = toolBarState[this.command].disabled;
    }
}
