// @progress/kendo-angular-listbox/esm2015/constants.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
/**
 * @hidden
 */
export const DEFAULT_TOOLBAR_POSITION = 'right';
/**
 * @hidden
 */
export const allTools = [
    {
        name: 'moveUp',
        label: 'Move Up',
        icon: 'arrow-n'
    },
    {
        name: 'moveDown',
        label: 'Move Down',
        icon: 'arrow-s'
    },
    {
        name: 'transferTo',
        label: 'Transfer From',
        icon: 'arrow-w'
    },
    {
        name: 'transferFrom',
        label: 'Transfer To',
        icon: 'arrow-e'
    },
    {
        name: 'transferAllTo',
        label: 'Transfer All To',
        icon: 'arrow-double-60-right'
    },
    {
        name: 'transferAllFrom',
        label: 'Transfer All From',
        icon: 'arrow-double-60-left'
    },
    {
        name: 'remove',
        label: 'Remove',
        icon: 'x'
    }
];
/**
 * @hidden
 */
export const sizeClassMap = {
    small: 'sm',
    medium: 'md',
    large: 'lg'
};
/**
 * @hidden
 */
export const toolbarClasses = {
    left: 'k-listbox-toolbar-left',
    right: 'k-listbox-toolbar-right',
    top: 'k-listbox-toolbar-top',
    bottom: 'k-listbox-toolbar-bottom'
};
