// @progress/kendo-angular-listbox/esm2015/selection.service.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { EventEmitter, Injectable } from "@angular/core";
import * as i0 from "@angular/core";
/**
 * @hidden
 */
export class ListBoxSelectionService {
    constructor() {
        this.onSelect = new EventEmitter();
        this.selectedIndex = null;
    }
    select(index) {
        this.selectedIndex = index;
        this.onSelect.next({ index: this.selectedIndex });
    }
    isSelected(index) {
        return index === this.selectedIndex;
    }
    clearSelection() {
        this.selectedIndex = null;
    }
}
ListBoxSelectionService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: ListBoxSelectionService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
ListBoxSelectionService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: ListBoxSelectionService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: ListBoxSelectionService, decorators: [{
            type: Injectable
        }] });
